### PR TITLE
Fix arithmetic operations issues for Numpy arrays.

### PR DESCRIPTION
**Pull request by Pinkyboi**
Fixes: [#1105](https://github.com/pyccel/pyccel/issues/1105) and [#1135](https://github.com/pyccel/pyccel/issues/1135)

- Make Pyccel prioritize Numpy arrays types in arithmetic operations.
- Avoid precision loss when performing binary operations on variables having different types.
- Create temporary variable holding the value of arithmetic operations on array for function arguments.
- Fix name clash error occurring when doing an arithmetic operation between an array and function's return.

### CODE REVIEW

**Comment by EmilyBourne** [comment]
> @EmilyBourne i think what i changed could fix both issues, but i wanted to know your opinion on the case of an `AugAssign` for example, how do you think that should be handled ?
> 
> Lets take for example one of the functions that fails during the tests:
> 
> ```python
> @types('int', 'int', 'int')
> @types('int8', 'int', 'int')
> @types('int16', 'int', 'int')
> @types('int32', 'int', 'int')
> @types('int64', 'int', 'int')
> @types('float', 'int', 'int')
> @types('float32', 'int', 'int')
> @types('float64', 'int', 'int')
> def get_linspace(start, steps, num):
>     stop = start + steps
>     b = linspace(start, stop, num)
>     x = 0.0
>     for bi in b:
>         x += bi
>     return x
> ```
> 
> We will get this error message:
> 
> ```
> ERROR at annotation (semantic) stage
> pyccel:
>  |error [semantic]: main.py [18,8]| Variable has already been defined with type 'float'. Type 'float32' in assignment is incompatible (x=bi)
> ```
> 
> That comes from one of my changes that prioritize the numpy type precision over the pure python type precision, is there a way to know if the lhs is already a defined variable so we can adapt the used precision to it instead of calculating it ?

If I run a quick test then I get:
```
In [1]: import numpy as np                                                      

In [2]: a = np.float32(4.5)                                                     

In [3]: b = 0.0                                                                 

In [4]: type(b)                                                                 
Out[4]: float

In [5]: type(a)                                                                 
Out[5]: numpy.float32

In [6]: b+= a                                                                   

In [7]: type(b)                                                                 
Out[7]: numpy.float64
```

So the error sounds right. I think it is the test that needs correcting. E.g:
```
def get_linspace(start, steps, num):
    stop = start + steps
    b = linspace(start, stop, num)
    x = np.float64(0.0)
    for bi in b:
        x += bi
    return x
```

**Comment by Pinkyboi** [comment]
@EmilyBourne  That's what i thought too, ill change those tests.

File: tests/errors/semantic/non_blocking/INCOMPATIBLE_TYPES_IN_ASSIGNMENT_2.py
Comment (EmilyBourne) [inline tests/errors/semantic/non_blocking/INCOMPATIBLE_TYPES_IN_ASSIGNMENT_2.py:5]: Can you explain this change

File: tests/epyccel/test_arrays.py
Comment (EmilyBourne) [inline tests/epyccel/test_arrays.py:4002]: You should also check the return type (maybe this explains your precision problems). If the type is wrong then this problem might be linked to an issue caught by @aihya and fixed in PR #1169

Comment (Pinkyboi) [reply tests/epyccel/test_arrays.py:4016]: If i understand correctly what @aihya explained to me is that the error could occur when converting an `np.array` to a python array the precision may go down in some cases, but i'm still not entirely sure about the implications of this.

**Comment by Pinkyboi** [comment]
@EmilyBourne It seems like the you are right the precision become lower in windows, should add the fix that @aihya added directly to my branche or should i leave it as it is ?

**Comment by EmilyBourne** [comment]
> @EmilyBourne It seems like the you are right the precision become lower in windows, should add the fix that @aihya added directly to my branche or should i leave it as it is ?

It's up to you and depends on how comfortable you are with git merges.

If you are comfortable then I would apply the fix locally to check that the problem is caused by this bug and not by pyccel assigning the wrong precision internally.

If you are not comfortable, then maybe wait until @yguclu has merged #1169

**Comment by Pinkyboi** [comment]
@EmilyBourne before adding a valid test for #1105  that would check if we prioritized the `numpy` type over the pure python type, i wanted to be sure that this behavior is pyccel specific since this implementation doesn't mimic the python implementation.
In case of the python implementation the the precision would get promoted if the initial precision is lower than the pure python type precision.

**Comment by EmilyBourne** [comment]
> @EmilyBourne before adding a valid test for #1105 that would check if we prioritized the `numpy` type over the pure python type, i wanted to be sure that this behavior is pyccel specific since this implementation doesn't mimic the python implementation. In case of the python implementation the the precision would get promoted if the initial precision is lower than the pure python type precision.

What exactly doesn't mimic the python implementation? The example provided in #1105?
What version of python do you have? On my computer (python 3.8.10) the precision does not get promoted:
```python
In [2]: import numpy as np 
   ...:  
   ...: def get_int_array(): 
   ...:     a = np.ones(5, dtype=np.int32) 
   ...:     b = a*2 
   ...:     return b[0] 
   ...:                                                                         

In [3]: type(get_int_array())                                                   
Out[3]: numpy.int32
```

**Comment by Pinkyboi** [comment]
@EmilyBourne I just got confused since the treatment on the scalar precision and array precision isn't the same.

File: pyccel/ast/operators.py
Comment (ohachim) [inline pyccel/ast/operators.py:395]: Will it ever pass to this line?

File: pyccel/ast/operators.py
Comment (ohachim) [inline pyccel/ast/operators.py:389]: Are the list comprehension necessary, it doesn't look like they'll be used later, wouldn't it be better to just use `args[0].dtype`?

**Review by ohachim — CHANGES_REQUESTED** [review]

File: pyccel/ast/operators.py
Comment (azouiten) [inline pyccel/ast/operators.py:388]: new line?

File: pyccel/ast/internals.py
Comment (azouiten) [inline pyccel/ast/internals.py:299]: I think these two statments are the same, so maybe you should just remove the if else?

File: tests/epyccel/recognised_functions/test_numpy_types.py
Comment (azouiten) [inline tests/epyccel/recognised_functions/test_numpy_types.py:18]: is there a reason why you import numpy in every function def?, I ve noticed that most tests have the same import statments.

File: tests/macro/scripts/MPI/gather.py
Comment (azouiten) [inline tests/macro/scripts/MPI/gather.py:19]: I remember emily once said that we should not make norm changes.

**Review by azouiten — DISMISSED** [review]

File: tests/macro/scripts/MPI/gather.py
Comment (EmilyBourne) [reply tests/macro/scripts/MPI/gather.py]: It's not that you shouldn't make norm changes. It's just that it makes it harder to review a PR when there are things in it that aren't relevant. It gives the impression that this PR changes something in MPI when it doesn't.
It's mostly courtesy. There's no need for all 3 reviewers to double check a file that didn't really change.

If we want to make changes to eg normalise the spacing conventions then we could create a PR for this and add another linter to keep it clean but this isn't a priority as the readability isn't very different and it doesn't affect the execution

**Comment by EmilyBourne** [comment]
@Pinkyboi I see that you have re-requested reviews from @ohachim and @azouiten , but you don't seem to have addressed most of their comments. Please leave a message on open comments either linking to the commit where you changed something or explaining why the change is unnecessary/can't be done.

**Comment by Pinkyboi** [comment]
@EmilyBourne it seems like i forgot to submit my comments, i didn't notice the the pending tag on my comments.

File: pyccel/ast/operators.py
Comment (EmilyBourne) [reply pyccel/ast/operators.py:394]: @ohachim good catch. I suspect that's my error originally.
@Pinkyboi would you mind changing the order of these lines

File: pyccel/ast/operators.py
Comment (EmilyBourne) [reply pyccel/ast/operators.py]: For example `3 + 4.0` will lead to a float with the current implementation, but would give an integer with your suggestion

File: tests/epyccel/test_arrays.py
Comment (EmilyBourne) [reply tests/epyccel/test_arrays.py]: @Pinkyboi Did you update the tests?

File: pyccel/ast/operators.py
Comment (Pinkyboi) [reply pyccel/ast/operators.py:394]: @EmilyBourne wouldn't it be better to leave just the `TypeError` raised by  `_handle_str_type` ? in my opinion it's more descriptive of the error than just an `assert` although the error message `"unsupported operand type(s) for /: 'str' and 'str'"` could maybe be changed.

Comment (EmilyBourne) [reply pyccel/ast/operators.py:394]: Ah if there is a `TypeError` then yes that is much better. In that case the `assert` can indeed be removed.
> `"unsupported operand type(s) for /: 'str' and 'str'"` could maybe be changed.

This doesn't seem like it's reporting the same error though, so we might need to add an additional `TypeError`. This is talking about doing `'ha'/'out'` which doesn't mean anything in python. Some of the other operators are supported but I think in that case they have their own classes which are not operators (e.g `Concatenate`).

File: tests/epyccel/test_arrays.py
Comment (Pinkyboi) [reply tests/epyccel/test_arrays.py]: @EmilyBourne i did revert back some changes there since i noticed that the problem was caused by my implementation, i left the test on the return type cuz i thought i may be useful still.

File: tests/epyccel/recognised_functions/test_numpy_funcs.py
Comment (Pinkyboi) [reply tests/epyccel/recognised_functions/test_numpy_funcs.py:5275]: @EmilyBourne if i remember correctly there was a problem with fortran when doing binary arithmetic operations on the fly,  for `fortran`, `merge()` if we replaced the test with something like : 
```python
    def get_chosen_elements(arr):
        from numpy import where, shape
        
        arr1 = arr + 2
        a = where(arr > 5, arr, arr1)
        s = shape(a)
        return len(s), s[0], a[1], a[0]
```  
I wasn't sure if this is within the scope of my issue to be honest

Comment (EmilyBourne) [reply tests/epyccel/recognised_functions/test_numpy_funcs.py:5275]: Either a temporary can be created for this, or you should use the cast functions if they handle arrays

**Review by ohachim — APPROVED** [review]

File: pyccel/parser/semantic.py
Comment (EmilyBourne) [inline pyccel/parser/semantic.py:715]: Do you need to visit `FunctionCallArgument(var)`? The `var` object has a type/precision/rank/etc so doesn't it already have all necessary semantic behaviour?

File: pyccel/parser/semantic.py
Comment (EmilyBourne) [inline pyccel/parser/semantic.py:1402]: `self._additional_exprs[-1]` is already a list so upacking it into a list is pointless.
```python
add_exprs  = [*self._additional_exprs[-1]]
```
is equivalent to:
```python
add_exprs  = list(self._additional_exprs[-1])
```
or
```python
add_exprs  = self._additional_exprs[-1].copy()
```
But there is no need to make a copy here.

Also `pop()` returns the element that it is removing so this section could be written much more compactly as:
```python
loops = self._additional_exprs.pop() + loops
self._additional_exprs.append([])
```

Comment (Pinkyboi) [reply pyccel/parser/semantic.py:1402]: Sure that looks cleaner, I'll fix that.

**Review by aihya — APPROVED** [review]

**Comment by yguclu** [comment]
@aihya Please answer all the comments made by @EmilyBourne, pointing out which commit applies the requested change.

File: pyccel/parser/semantic.py
Comment (Pinkyboi) [reply pyccel/parser/semantic.py]: I ended up dropping the insertion in the loop list in [b26a937](https://github.com/pyccel/pyccel/pull/1187/commits/b26a937f05f32cee939135004c3977b37f9b462f).

File: pyccel/parser/semantic.py
Comment (Pinkyboi) [reply pyccel/parser/semantic.py]: i removed the visit to `FunctionCallArgument`  in [51fe564](https://github.com/pyccel/pyccel/pull/1187/commits/51fe56464f3b9dcc1d309dcd559434a087da4b05).

**Comment by Pinkyboi** [comment]
@yguclu Really sorry for that, i added comments pointing in which commit the changes occurred.

File: tests/epyccel/recognised_functions/test_numpy_funcs.py
Comment (yguclu) [reply tests/epyccel/recognised_functions/test_numpy_funcs.py]: @Pinkyboi Has this comment been addressed?

File: pyccel/ast/internals.py
Comment (EmilyBourne) [inline pyccel/ast/internals.py:298]: Is there a reason to prefer `attrgetter` to ` get_final_precision(max(ndarray_list, key=lambda l: l.precision))`?

File: pyccel/parser/semantic.py
Comment (EmilyBourne) [inline pyccel/parser/semantic.py:713]: Please add a test for this case

**Review by EmilyBourne — CHANGES_REQUESTED** [review]

File: pyccel/ast/internals.py
Comment (Pinkyboi) [reply pyccel/ast/internals.py:298]: I guess using the lambda function improves the readability a bit, i'll switch to that i guess.

Comment (EmilyBourne) [reply pyccel/ast/internals.py:298]: I don't know if it's more readable, but it removes an unnecessary import

Comment (EmilyBourne) [reply pyccel/ast/internals.py:298]: Actually it looks like your choice is faster: https://stackoverflow.com/a/54061395/5957195 :+1:

Comment (Pinkyboi) [reply pyccel/ast/internals.py:298]: Alright i'll leave it as it is.

File: tests/epyccel/recognised_functions/test_numpy_funcs.py
Comment (Pinkyboi) [reply tests/epyccel/recognised_functions/test_numpy_funcs.py]: @yguclu since all the commits after this message aim to fix that issue, i wasn't sure if i was supposed to address it in any way.

File: pyccel/parser/semantic.py
Comment (Pinkyboi) [reply pyccel/parser/semantic.py:713]: I added some tests in [2158507](https://github.com/pyccel/pyccel/pull/1187/commits/215850701c290467e536c5c1bae4755cc9b1c38a), is it alright to add the tests in scripts instead of epyccel, because i wanted to add tests for arithmetic operations with functions return.

Comment (EmilyBourne) [reply pyccel/parser/semantic.py:713]: Tests in the `tests/pyccel/` folder are fine, but you need to add them to the `test_pyccel.py` file in order for them to be executed. I don't see any tests testing this though. Is it not supposed to handle things like:
```
f(a+b)
```
?

Comment (Pinkyboi) [reply pyccel/parser/semantic.py:713]: Oh i guess i didnt add em, i was testing em in already defined functions like in `np.array` is it fine ?

Comment (Pinkyboi) [reply pyccel/parser/semantic.py:713]: @EmilyBourne i did add a test with an arithmetic add in [fdf0609](https://github.com/pyccel/pyccel/pull/1187/commits/fdf06091e95635660e96f393a40e239bf52e2805).

Comment (EmilyBourne) [reply pyccel/parser/semantic.py:713]: > Oh i guess i didnt add em, i was testing em in already defined functions like in np.array is it fine ?

I don't see this test. Can you point me to the line

Comment (EmilyBourne) [reply pyccel/parser/semantic.py:713]: Never mind I found it. For some reason I thought this code only applied to user defined functions

File: tests/epyccel/recognised_functions/test_numpy_types.py
Comment (EmilyBourne) [inline tests/epyccel/recognised_functions/test_numpy_types.py:81]: It seems a shame not to also test the result in this test function

Comment (Pinkyboi) [reply tests/epyccel/recognised_functions/test_numpy_types.py:81]: Surely, i will add that.

Comment (Pinkyboi) [reply tests/epyccel/recognised_functions/test_numpy_types.py:87]: Done in [931707b](https://github.com/pyccel/pyccel/pull/1187/commits/931707bcf90417a380050074300fa42212e35785).

**Review by EmilyBourne — APPROVED** [review]

**Review by yguclu — APPROVED** [review]

**Pull request merged**
